examples/jsm/renderers/webgpu/WebGPURenderObject.js

```diff
@@ -0,0 +1,68 @@
+export default class WebGPURenderObject {
+
+	constructor( renderer, properties, object, material, scene, camera, lightsNode ) {
+
+		this.renderer = renderer;
+		this.properties = properties;
+		this.object = object;
+		this.material = material;
+		this.scene = scene;
+		this.camera = camera;
+		this.lightsNode = lightsNode;
+
+		this.geometry = object.geometry;
+
+		this._materialVersion = - 1;
+		this._materialCacheKey = '';
+
+	}
+
+	getEnvironmentNode() {
```

Comment:
The PR looks great so far! The only thing that I have noticed is that some node-specific logic/semantic finds its way into the renderer.

The original idea was to make the renderer material-agnostic (as good as possible) so it does not rely on a specific material system. But now there are things like `getEnvironmentNode()` or `getFogNode()` (so node syntax) inside renderer classes.

Would it be possible to reduce the node related logic inside `WebGPURenderer` to a minimum? 

The idea would be to define more general interfaces in the renderer (e.g. `getEnvironment()`) and then inject a material specific object (something like `nodes` that we already used elsewhere) that provides the implementation. In our case it is node specific but in theory it could be something different, too.

The overall design goal is to avoid a too tight coupling of the renderer to the node material.

Reply:
I think it ideally should be something like this:
WebGLRenderer/WebGPURenderer/SoftwareRenderer/another renderer (doesn't know about nodes, takes shaders written in GLSL/WGSL/JS) -- NodeRenderer (a boilerplate "renderer" that knows how to work with nodes) and WebGLNodeBuilder/WebGPUNodeBuilder/JSNodeBuilder (converts node material to GLSL/WGSL/JS shader) -- NodeMaterial system -- Nodes system.
So here `getEnvironmentNode` and `getFogNode` could go exactly in the NodeRenderer.

Reply:
Current goals are these:

1. `WebGPURenderer` should be able to replace `WebGLRenderer` (except `ShaderMaterial`).
2. `WebGPURenderer` should have a `WebGPUBackend` and a WebGLBackend`.
3. This summer, the `<model-viewer>` project has to be able to replace `WebGLRenderer` with `WebGPURenderer`  easily so it can use WebGPU when available.
4. After replacing the renderer, `<model-viewer>`'s file size should stay relatively the same.

We should be careful with any additional work on the nodes system that gets in the way of achieving these goals.

Reply:
Why we need WebGL backend for _WebGPU_Renderer? Why we can't make (almost) completely interchangeable WebGLRenderer and WebGPURenderer?

Reply:
Because we can't ask developers to start using two renderers in their projects and do the feature detection themselves. Resulting in 2x bigger build sizes too.

The original goal of the nodes code was to solve the tree-shaking problem of the `WebGLRenderer` by replacing the non tree-shakeable `ShaderChunks` with the tree-shakeable `NodeMaterial`.

The fact that `NodeMaterial` is able to produce GLSL and WGSL makes it much easier to have the renderer support both back-ends..

And developers should not have to know any of these details.

It seems to me like the best plan is to say that... as long as their project doesn't use `ShaderMaterial`... we can promise them that they can start using `WebGPURenderer` in their projects if they want to get WebGPU performance benefits when available, yet the renderer will take care of the backwards compatibility for them when not available.

So, the community's developer experience I'm aiming for is this:

1. Same scene graph API.
2. The only thing that changes is the renderer.
3. Their projects will automatically support WebGL and WebGPU.
4. If `NodeMaterial` is able to keep its promise, their bundle sizes should be smaller.

We may need to create a separate npm package though... `three-webgpu` or something.

Reply:
> Because we can't ask developers to start using two renderers in their projects and do the feature detection themselves.

It's one line of code. `const renderer = WebGPU.isAvailable() ? new WebGPURenderer() : new WebGLRenderer`. We already make developers to do various things by hand, e.g. auto-changing aspect ratio when resizing -- why this can't be one more thing?

> Resulting in 2x bigger build sizes too.

Wouldn't WebGL and WebGPU backends result in almost the same increase? Also, a very significant part of the build is renderer-independent (like animations) or shader-related (which could be repalced by nodes and then shared by both renderers).

> And developers should not have to know any of these details.

Developers **should** know if their application uses WebGL or WebGPU. The application shouldn't automatically try to use WebGL if told to use WebGPU and vice versa.

It's more about naming -- I just say that we shouldn't go with WebGLBackend and WebGPUBackend of WebGPURenderer, we should rather stick to WebGLRenderer and WebGPURenderer.
Even creating something like TheRenderer that would have WebGL and WebGPU as backends would be less confusing.

Reply:
The WebGL to WebGPU transition strategy I've just described is the strategy I feel I can take responsibility for and has the least negative impact in the community and ecosystem.

Reply:
I think having a universal renderer can help in many ways,one of the difficulties I have in porting NodeMaterial to WebGLRenderer is because it wasn't exactly developed for that purpose, so it still contains many hacks, and many features that we see today in WebGPURenderer we just don't see in WebGLRenderer because the implementation is very difficult without modifying the core, and would probably impact legacy projects.  I fell TSL will help reduce duplicate code. Currently, I'm working on backward compatible  and stabilizing the WebGPURenderer, but if we don't have a universal renderer like `THREE.Renderer()` it would be difficult to guarantee the compatibility of new features of WebGPURenderer like LightNode API ( IBL's and CustomLights ), NodeMaterial Particles, Skinning Instacing for example to WebGLRenderer.

Reply:
I think this is exactly what I meant by NodeRenderer [here](https://github.com/mrdoob/three.js/pull/25750#discussion_r1155761766).

Reply:
@sunag 

> I think having a universal renderer can help in many ways,one of the difficulties I have in porting NodeMaterial to WebGLRenderer is because it wasn't exactly developed for that purpose, so it still contains many hacks, and many features that we see today in WebGPURenderer we just don't see in WebGLRenderer because the implementation is very difficult without modifying the core, and would probably impact legacy projects.

How about making it so `NodeMaterial` only works in `WebGPURenderer`?

Reply:
> That line of code explodes the amount of reasons why something may not work in some random device/browser configurations.

A `const renderer = new WebGPURenderer()`, if WebGPURenderer will try to autodetect and use WebGL or WebGPU backend, will explode no less.

> If NodeMaterial has been designed well, only shaders that the developer's project use will be included when producing a bundle. That's something we can't fix anymore with WebGLRenderer but WebGPURenderer gives us a chance to make it right for the next generation.

I mean that bundling something like WebGLNodeRenderer and WebGPURenderer would be almost the same file size-wise as bundling WebGPURenderer with WebGLBackend and WebGPUBackend.

> But you're thinking about advanced developers, which may just be 10% of Three.js users. 90% of Three.js users have never written a shader.

I don't think it's too advanced if a developer just wants to create WebGL-exclusive or WebGPU-exclusive program.

> If we name it THREE.Renderer and in the future there's a new API, say WebRaytracing... then we will be forced to bloat THREE.Renderer.

> WebGPU gives us a chance to clean up the scene graph API a bit. We do not have to make WebGPURenderer fully compatible, but we do have to make a WebGL back-end for it, otherwise no one will use it because "it has not reached 99% support yet".

This is again a reason why separate renderers are better. If in future we will have, for example, WebRaytracingRenderer that tries to use WebRaytracing or WebGPU and WebGPURenderer that tries to use WebGPU or WebGL... It would be a terrible mess.

Feature detection via `WebGPU.isAvailable()` isn't too hard even for a non-advanced developer to do. At least the developer would understand then why their WebGL application breaks in WebGPU or vice versa.

Reply:
I think we could consider that possibility:

```js
const backend = WebGPU.isAvailable() ? new WebGPUBackend() : new WebGLBackend();

const renderer = new THREE.Renderer( backend );
```

Reply:
Perhaps if the backend was compatible with Raytracing as well like:
But I don't know if it would be possible, just a thought
```js
const renderer = new THREE.Renderer( new WebGPURaytracingBackend() );
```

Reply:
> `const backend = WebGPU.isAvailable() ? new WebGPUBackend() : new WebGLBackend(); const renderer = new THREE.Renderer( backend );`

That's acceptable, I think -- but `const renderer = WebGPU.isAvailable() ? new WebGPURenderer() : new WebGLNodeRenderer()` is semantically the same but little bit simpler. But I don't oppose this approach.

Reply:
That's basically what the same the `WebGPURenderer` will do internally.

```js
const backend = WebGPU.isAvailable() ? new WebGPUBackend() : new WebGLBackend();
```